src/common/tusb_fifo.h

```diff
@@ -134,8 +142,10 @@ void     tu_fifo_advance_read_pointer   (tu_fifo_t *f, uint16_t n);
 // This functions deliver a pointer to start reading/writing from/to and a valid linear length along which no wrap occurs.
 // In case not all of your data is available within one read/write, update the read/write pointer by
 // tu_fifo_advance_read_pointer()/tu_fifo_advance_write_pointer and conduct a second read/write operation
-uint16_t tu_fifo_get_linear_read_info   (tu_fifo_t *f, uint16_t offset, void **ptr, uint16_t n);
-uint16_t tu_fifo_get_linear_write_info  (tu_fifo_t *f, uint16_t offset, void **ptr, uint16_t n);
+// TODO - update comments
+
+tu_fifo_linear_wr_info tu_fifo_get_linear_read_info(tu_fifo_t *f, uint16_t offset, uint16_t n);
```

Comment:
1. Do we ever need to use `offset`. If not I think we can drop that parameter for a simpler code.
2. I think we could just drop `n` as well, just return len1 and len2 as it is available from buffer. This API would be like query the current buffer state of the fifo. And we could rename function to  `tu_fifo_get_read_info()` or `tu_fifo_get_read_buffer_info()`
3. As general of thumb, we shouldn't return a struct object (> 4 bytes), it should be passed from application to fill out data. Function return can be `bool`. Return false when there is not 

Similar comment apply to `tu_fifo_get_linear_write_info`

Reply:
I would go with `tu_fifo_get_read_info()`.
To 3: Where does this restriction come from? And return `false` when there is no place to write to or nothing to read from?

Reply:
I don't know if offset will be needed or when the offset came up the first time... we can remove it completely from the `fifo.c`, i.e. also from the write and read functions.

Reply:
> To 3: Where does this restriction come from?

Depends on the compiler, sometimes it will make a copy of `tu_fifo_buffer_info_t info`.

Reply:
> I would go with `tu_fifo_get_read_info()`.
> To 3: Where does this restriction come from? And return `false` when there is no place to write to or nothing to read from?

Yeah right, I was mixing thought, I think we can have it to return void if both offset and n are dropped from the argument as 1 and 2

Reply:
> I don't know if offset will be needed or when the offset came up the first time... we can remove it completely from the `fifo.c`, i.e. also from the write and read functions.

Yeah, I think offset is is kind of needed for peek_at() initialy, which is trivial with initial implementation. Though If needed, we can just removed the peek_at(offset), and just implement peek() to look at current index only. peek_at random offset is not used by any class driver/application so far. So we don't really need it :) and clean API is much better needed.

Reply:
Ok i have but one question, the first time i implemented the `get_read_info()` i did it without struct but with poiners `get_read_info(* uint16_t len_1, * uint16_t len_2, ...)`. However, the CI compilers complained, that `len_1` etc. were potentially uninitialized and aborted with an error. To fix this, i came up with the struct at the very end. Now, if i change it to `get_read_info(* buffer_info into)`, then i bet the compiler complains again. This would force the user to always init the struct with zeros or so. Is there a way to tell the compiler that the struct will hold valid values in any case?

Reply:
hmm, that is weird, did you have any branch in the function that doesn't assign value for the struct memeber. Like invalid short return, it would cause gcc to trigger possible path for uninitialized value.

Reply:
Me yes... in case there was nothing to read i set `len_1 = 0` and left the rest uninitialized to save time.
Nevermind, i change it to the pointer version to the struct if this is your desired design. I guess its not worth spending more time on this ^^

Reply:
Yes, please do. I do prefer it that way .